source/slang/ir-legalize-types.cpp

```diff
@@ -251,6 +440,442 @@ static bool isResourceType(Type* type)
+        }
+
+        if (specialType.flavor != LegalType::Flavor::none)
+        {
+            anySpecial = true;
+            anyComplex = true;
+            pairElement.flags |= PairInfo::kFlag_hasSpecial;
+
+            TuplePseudoType::Element specialElement;
+            specialElement.fieldDeclRef = fieldDeclRef;
+            specialElement.type = specialType;
+            specialElements.Add(specialElement);
+        }
+
+        pairElement.type = LegalType::pair(ordinaryType, specialType, elementPairInfo);
+        pairElements.Add(pairElement);
+    }
+
+    // Add a field to the (pseudo-)type we are building
+    void addField(
+        DeclRef<VarDeclBase>    fieldDeclRef)
+    {
+        // Skip `static` fields.
+        if (fieldDeclRef.getDecl()->HasModifier<HLSLStaticModifier>())
+            return;
+
+        auto fieldType = GetType(fieldDeclRef);
+
+        bool isResourceField = isResourceType(fieldType);
+
+        auto legalFieldType = legalizeType(context, fieldType);
+        addField(
+            fieldDeclRef,
+            legalFieldType,
+            legalFieldType,
+            isResourceField);
+    }
+
+    LegalType getResult()
+    {
+        // If we didn't see anything "special"
+        // then we can use the type as-is.
+        // we can conceivably just use the type as-is
+        //
+        // TODO: this might be a good place to turn
+        // a reference to a generic `struct` type into
+        // a concrete non-generic type so that downstream
+        // codegen doesn't have to deal with generics...
+        //
+        // TODO: In fact, why not just fully replace
+        // all aggregate types here with some structural
+        // types defined in the IR?
+        if (!anyComplex)
+        {
+            return LegalType::simple(type);
+        }
+
+        // If there were any "ordinary" fields along the way,
+        // then we need to collect them into a type to
+        // represent the ordinary part of things.
+        //
+        RefPtr<Type> ordinaryType;
+        if (anyOrdinary)
+        {
+            RefPtr<FilteredTupleType> ordinaryTypeImpl = new FilteredTupleType();
+            ordinaryTypeImpl->setSession(context->session);
+            ordinaryTypeImpl->originalType = type;
+            ordinaryTypeImpl->elements = ordinaryElements;
+            ordinaryType = ordinaryTypeImpl;
+        }
+
+        LegalType specialType;
+        if (anySpecial)
+        {
+            RefPtr<TuplePseudoType> specialTuple = new TuplePseudoType();
+            specialTuple->elements = specialElements;
+            specialType = LegalType::tuple(specialTuple);
+        }
+
+        RefPtr<PairInfo> pairInfo;
+        if (anyOrdinary && anySpecial)
+        {
+            pairInfo = new PairInfo();
+            pairInfo->elements = pairElements;
+        }
+
+        return LegalType::pair(ordinaryType, specialType, pairInfo);
+    }
+
+};
+
+static RefPtr<Type> createBuitlinGenericType(
```

Comment:
A typo here?

Reply:
Good catch. Will fix this in the same commit where I try to fix the test failures.